rs/src/parse/traversal.rs

```diff
@@ -584,33 +584,14 @@ pub fn parse_proto<T, F, B>(
     root_parser: F,
     state: &mut context::State,
     config: &config::Config,
-) -> parse_result::ParseResult
+) -> diagnostic::Result<parse_result::ParseResult>
 where
     T: prost::Message + InputNode + Default,
     F: FnOnce(&T, &mut context::Context) -> diagnostic::Result<()>,
     B: prost::bytes::Buf,
 {
     match T::decode(buffer) {
-        Err(err) => {
-            // Create a minimal root node with just the decode error
-            // diagnostic.
-            let mut root = T::type_to_node();
-
-            // Create a root context for it.
-            let mut context = context::Context::new(root_name, &mut root, state, config);
-
-            // Push the diagnostic using the context.
-            context.push_diagnostic(diagnostic::RawDiagnostic {
-                cause: ecause!(ProtoParseFailed, err),
-                level: diagnostic::Level::Error,
-                path: path::PathBuf {
-                    root: root_name,
-                    elements: vec![],
-                },
-            });
-
-            parse_result::ParseResult { root }
-        }
+        Err(err) => Err(ecause!(ProtoParseFailed, err)),
```

Comment:
Nit: you can probably use `.map_err` and `map` with `?` here.

Reply:
I didn't go full functional by applying `map` for the body of the `Ok` branch, but still simplified it a great deal in 0932ed5.